### PR TITLE
Disable implicit casts

### PR DESCRIPTION
**Pull request by natebosch**
This will make it easier to migrate to null safety which also disables
implicit casts except from dynamic. Also fix the implicit casts from
dynamic since this package has a lot of dynamic code that is harder to
read with implicit casts.

- Add argument types to some signatures that should have had them from
  the start.
- Add explicit casts.
- Extract a few new local variables to make some of the casts only need
  to happen once.
- Use `Iterable.whereType` in a loop instead of a type check and
  `continue` in the loop body.

### CODE REVIEW

**Comment by natebosch** [comment]
As an FYI I also plan on enabling `prefer_final_locals` because I'm finding that it make codes written like this (overlong functions etc) easier to read.

File: lib/dom.dart [hunk 1]
```diff
@@ -1058,14 +1062,14 @@ class FilteredElementList extends IterableBase<Element>
   // bug where ListMixin<E>.indexOf() expects Object as the argument.
   @override
   int indexOf(Object element, [int start = 0]) =>
-      _filtered.indexOf(element, start);
+      _filtered.indexOf(element as Element, start);
```
Comment (jakemac53) [inline lib/dom.dart:1065]: Is the above TODO still an issue or can we type the arg as Element? (ditto for next one)

File: lib/parser.dart [hunk 2]
```diff
@@ -1816,7 +1818,7 @@ class InBodyPhase extends Phase {
     if (tree.formPointer != null) {
       return;
     }
-    var formAttrs = <dynamic, String>{};
+    var formAttrs = LinkedHashMap<dynamic, String>();
```
Comment (jakemac53) [inline lib/parser.dart:1821]: Why this change?

File: lib/src/query_selector.dart [hunk 3]
```diff
@@ -40,21 +40,19 @@ class SelectorEvaluator extends Visitor {
   }
 
   Element querySelector(Node root, SelectorGroup selector) {
-    for (var node in root.nodes) {
-      if (node is! Element) continue;
-      if (matches(node, selector)) return node;
-      var result = querySelector(node, selector);
+    for (var element in root.nodes.whereType<Element>()) {
```
Comment (jakemac53) [inline lib/src/query_selector.dart:43]: Probably doesn't matter but the previous implementation could have been done for speed reasons?

**Review by jakemac53 — APPROVED** [review]

File: lib/dom.dart (hunk 1 above)
Comment (natebosch) [reply lib/dom.dart:1065]: Still an issue and likely wont' be changed for a while. Would you prefer I remove the comments or update them to point to https://github.com/dart-lang/sdk/issues/31311

File: lib/parser.dart (hunk 2 above)
Comment (natebosch) [reply lib/parser.dart:1821]: It needs to be typed as `LinkedHashMap` below.

https://github.com/dart-lang/html/blob/25806e5e8e30f68aee4f108f229a22d9706ed589/lib/src/token.dart#L25

File: lib/src/query_selector.dart (hunk 3 above)
Comment (natebosch) [reply lib/src/query_selector.dart:43]: When it was first written we didn't have `whereType` as an option. I don't _think_ the performance is going to matter much here, so I preferred to use this over introducing yet another local variable or adding multiple casts.

File: lib/dom.dart (hunk 1 above)
Comment (jakemac53) [reply lib/dom.dart:1065]: Seems like just remove the TODO as it is unlikely to get resolved based on that issue

Comment (natebosch) [reply lib/dom.dart:1064]: I added a comment so that it's clear why there is a cast, but it's not marked as a TODO anymore.
